eclair-core/src/main/scala/fr/acinq/eclair/db/sqlite/SqliteAuditDb.scala

```diff
@@ -175,61 +208,94 @@ class SqliteAuditDb(sqlite: Connection) extends AuditDb with Logging {
     }
 
   override def listSent(from: Long, to: Long): Seq[PaymentSent] =
-    using(sqlite.prepareStatement("SELECT * FROM sent WHERE timestamp >= ? AND timestamp < ? ORDER BY timestamp")) { statement =>
+    using(sqlite.prepareStatement("SELECT * FROM sent WHERE timestamp >= ? AND timestamp < ?")) { statement =>
       statement.setLong(1, from)
       statement.setLong(2, to)
       val rs = statement.executeQuery()
-      var q: Queue[PaymentSent] = Queue()
+      var sentByParentId = Map.empty[UUID, PaymentSent]
       while (rs.next()) {
-        q = q :+ PaymentSent(
-          UUID.fromString(rs.getString("id")),
-          rs.getByteVector32("payment_hash"),
-          rs.getByteVector32("payment_preimage"),
-          Seq(PaymentSent.PartialPayment(
-            UUID.fromString(rs.getString("id")),
-            MilliSatoshi(rs.getLong("amount_msat")),
-            MilliSatoshi(rs.getLong("fees_msat")),
-            rs.getByteVector32("to_channel_id"),
-            None, // we don't store the route
-            rs.getLong("timestamp"))))
+        val parentId = UUID.fromString(rs.getString("parent_payment_id"))
+        val part = PaymentSent.PartialPayment(
+          UUID.fromString(rs.getString("payment_id")),
+          MilliSatoshi(rs.getLong("amount_msat")),
+          MilliSatoshi(rs.getLong("fees_msat")),
+          rs.getByteVector32("to_channel_id"),
+          None, // we don't store the route
+          rs.getLong("timestamp"))
+        val sent = sentByParentId.get(parentId) match {
+          case Some(s) => s.copy(parts = s.parts :+ part)
+          case None => PaymentSent(
+            parentId,
+            rs.getByteVector32("payment_hash"),
+            rs.getByteVector32("payment_preimage"),
+            MilliSatoshi(rs.getLong("recipient_amount_msat")),
+            PublicKey(rs.getByteVector("recipient_node_id")),
+            Seq(part))
+        }
+        sentByParentId = sentByParentId + (parentId -> sent)
       }
-      q
+      sentByParentId.values.toSeq.sortBy(_.timestamp)
```

Comment:
Wonder if using two collections would help (one to index by parent id, one to manage ordering).

Reply:
Sorting is considered a "for-free" primitive, I think it's less expensive to sort than to duplicate the amount of memory used. If we see a bottleneck there we can start experimenting.